TELCODOCS-213 D/S Doc: MGMT-3573, Assisted service support installing single node OCP - 2 (#33982)
openshift/openshift-docs: modules/ztp-ai-install-ocp-clusters-on-bare-metal.adoc

> // TELCODOCS-213 4.8 Using Assisted Installer to Install & Manage ocp Clusters on Bare Metal

Reviewer: Ad required preamble: https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#module-file-metadata

Author: Updated preamble as suggested.

//
//
//

[id=ztp-ai-install-ocp-clusters-on-bare-metal_{context}"]

.Enabling Assisted Installer Service on Bare Metal

> The Assisted Installer Service (AIS) deploys {product-title} clusters. Red Hat Advanced Cluster Management (RHACM) 2.3+ ships with AIS, which is deployed when the MultiClusterHub Operator is enabled on the RHACM hub cluster.

Reviewer: Don't use a hard coded product version, it will go out of date. Also, FYI, my PR adds two variables for ACM, `{rh-rhacm-first}`, which expands as  "Red Hat Advanced Cluster Management (RHACM)" and `{rh-rhacm}`, which expands as "ACM". I think you should replace and use these instead. My work is not merged yet, is the only problem.

> RHACM 2.3+ only supports *Single Node OpenShift (SNO)* cluster deployments. A SNO cluster runs OpenShift Container Platform (ocp) on top of a single BMO node. When the host is provisoned, the single node cluster acts as both supervisor and worker node.

Reviewer: What about: For distributed units DUs, {rh-rhacm} supports single node OpenShift deployments. Single node OpenShift runs on a single bare-metal host. When the host is provisoned, the single node cluster acts as both supervisor and worker node.

.Enable assisted installer service

.Prerequisites

> . OpenShift Container Platform 4.8+\{product-version}

Reviewer: . Install {product-title} {product-verstion} on a hub cluster.
. Install {rh-acm} and create the `MultiClusterHub` resource.

. https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.2/html/install/installing#installing-red-hat-advanced-cluster-management-from-the-operatorhub[RHACM 2.3+ with MultiClusterHub resource created]

> . `PersistentVolume` CRs for database and file system storage

Reviewer: Make it a complete sentence, parallelism in bullets where possible - "Create `PersistentVolume` CRs for database and file system storage".

Reviewer: Also, exand the first instance of CR as "custom resource (CR)", and there after refer to CRs.

> . Install `openshift-install` [cli tool] for the desired version of OCP

Reviewer: I don't think OCP is valid as a contraction for OpenShift prod name. Try: "Install `openshift-install` CLI for the desired version of {product-title}."

.Next Steps

. Modify `HiveConfig` to enable the feature gate for Assisted Installer
+
[source,bash]
----
 oc patch hiveconfig hive --type merge -p '{"spec":{"targetNamespace":"hive","logLevel":"debug","featureGates":{"custom":{"enabled":["AlphaAgentInstallStrategy"]},"featureSet":"Custom"}}}'
----

> 

Reviewer: s/Modify `provisioning`/Modify the `provisioning`

Author: Updated accordingly.

. Modify `provisioning` resource to allow the Bare Metal Operator to watch all namespaces
+
[source,bash]
----
 oc patch provisioning provisioning-configuration --type merge -p '{"spec":{"watchAllNamespaces": true }}'
----

. Create an `AgentServiceConfig` custom resource to install the Assisted Installer Service's controller
+
[source,bash]
----

>  cat <<EOF | oc create -f -

Reviewer: I don't think we should use these kind of cat <<EOF conveniences in the docs, but I could be wrong, my preference would be to have a seperate [source,yaml] block with just the YAML, and then a seperate command to apply or create the resource.

 apiVersion: agent-install.openshift.io/v1beta1
 kind: AgentServiceConfig
 metadata:
   name: agent
   namespace: assisted-installer
 spec:
   databaseStorage:

>      volumeName: <db-pv-name> <1>

Reviewer: nit: variables should use underscores, eg., <db_pv_name>

<4> Volume size of the `fs-pv-name` PersistentVolume.
<5> OCP version to boot the SNO clusters with.
<6> The rootfs ISO URL; part of step 4 of the prerequisites.
<7> The image url; part of step 4 of the prerequisites.

.Troubleshooting

.Managed cluster fails to come online

. Check the `AgentClusterInstall` install status:
[source,terminal]
----
oc get agentclusterinstall -n <cluster-name>
----

If under the `INSTALLED` column is `false`, the installation was unsuccessful. Use `describe` to view error details.

[source,terminal]
----
oc describe agentclusterinstall -n <cluster-name> <cluster-name>
----

. Amend the errors and do the following:

Remove the cluster's namespace.
[source,terminal]
----
 oc delete namespace <cluster-name>
----
This deletes all the namespace-scoped custom resources created for this SNO cluster.

. Remove the cluster's managed cluster resource.
[source,terminal]
----
 oc delete managedcluster <cluster-name>
----

. Redo the <<custom-resource-creation,Custom Resource Creation>> section for this cluster.

.Additional Resources

Kubernetes Custom Resources explained within the context of Assisted Installer Service.

> |===

Reviewer: The "New resource" column is not relevant and should be removed.

| Resource | Explanation | Usage | Scoping | New Resource

| BareMetalHost
| Contains the connection information for the BMC of the target Bare Metal machine
| Loads and boots the Discovery ISO on the target machine using the Redfish protocol.
| Namespace-scoped
| No*

| InfraEnv

> | Contains information for pulling OCP onto the target Bare Metal machine

Reviewer: replace OCP with {product-title} throughout.

| Used with ClusterDeployment to generate the Discovery ISO for the SNO cluster.
| Namespace-scoped
| Yes

| AgentClusterInstall

> | Specifies the SNO cluster's configuration such as networking, number of supervisor (control plane) nodes, etc. Will show the kubeconfig and credentials after completed installation.

Reviewer: replace SNO cluster with single node OpenShift throughout.

> | Used to specify the SNO cluster's configuration information and provide status during the installation of the cluster.

Reviewer: This is useful information, but I don't think it fits in a procedure. I would be inclined to rewite this content as an "overview of AI" concept module.

| Yes

| ManagedCluster
| When a cluster will be managed by the Hub, it needs to be imported and known. This Kubernetes object provides that interface.
| The Hub uses this resource to provision, manage and show status of clusters.
| Cluster-scoped
| No

| KlusterletAddonConfig
| Contains the list of services (provided by the Hub) to be deployed to a ManagedCluster
| Tells the Hub which addon services to deploy to a ManagedCluster
| Namespace-scoped
| No

| Namespace
| Logical space for each ManagedCluster's resources existing on the Hub
| Used to propagate resources to the ManagedCluster
| Cluster-scoped
| No

| Secret
| Three are created: BMC Secret, Private key Secret, and Image Pull Secret
| *BMC Secret* is used to authenticate into the target Bare Metal machine using its username and password. *Private key Secret* is used to SSH into the target Bare Metal machine. *Image Pull Secret* contains authentication information for the ocp image installed on the target Bare Metal machine.
| Namespace-scoped
| No

| ClusterImageSet
| Contains OCP image information such as the repository and image name
| Passed into resources to provide ocp images
| Cluster-scoped
| No
|===

> NOTE: *These resources are new within the context of RHACM, not ocp.

Reviewer: I think this note can be removed. also, FYI the note format we use is:

[NOTE]
====
note text
====

"ocp" or OCP are not used, use {product-title}

.Glossary

Definitions and explanations of abbreviations.

> [frame="topbot",options="header"]

Reviewer: I don't think this table is required.

// Module included in the following assemblies:
//

> // scalability_and_performance/ztp-deploying-disconnected.adoc

Reviewer 2: Are you going to include this module in an assembly? I'd love to be able to take a look at in in context and review the generated text.

Author: This PR is part of the `ztp-deploying-disconnected` assembly.

Reviewer 2: Then you need to add the preview link to the first comment. (You have to go back and edit the comment after Netlify renders it, and it's best to link as close to the relevant change that you're making as possible.)

Author: Acked.

[id="enabling-assisted-installer-service-on-bare-metal_{context}"]

> .Enabling Assisted Installer Service on Bare Metal

Reviewer 3: Wanted to verify if this is the final name?

Author: I believe so, but I'll follow up.

Author: Updated as suggested.

> .Enabling assisted installer service on bare metal

Reviewer 2: ```suggestion
Enabling assisted installer service on bare metal
```
This needs to be a heading, not a title.

Author: Updated as suggested.

The Assisted Installer Service (AIS) deploys {product-title} clusters. {rh-rhacm-first} ships with AIS, which is deployed when the MultiClusterHub Operator is enabled on the {rh-rhacm} hub cluster.

> For distributed units (DUs), {rh-rhacm} supports single node OpenShift deployments, which run on a single bare metal host. The single node cluster acts as both a control plane and a worker node. {rh-rhacm} supports single node OpenShift deployments for {product-title} on top of a single BMO DU.

Reviewer 2: Is Single Node OpenShift an approved service/entity name? You can't use bare "OpenShift" unless it's part of a discrete named service. The vast majority of the time, you need to use {product-title} to get the entire product name.

Please confirm the SNO usage and adjust, if required.

Also, what is the BMO? You need to expand this initialism on first use.

Author: Updated as suggested.

> .Enable assisted installer service

Reviewer 2: ```suggestion

```

Author: Deleted as suggested.

.Prerequisites

* Install {product-title} {product-verstion} on a hub cluster.

> * Install {rh-acm} and create the MultiClusterHub resource.

Reviewer 3: Some places use `{rh-rhacm}` but here `{rh-acm}`?

Author: Great catch! I'll follow up.

Author: Updated variables as suggested.

> * Install {rh-rhacm} and create the MultiClusterHub resource.

Reviewer 2: ```suggestion
* Install {rh-rhacm} and create the `MultiClusterHub` resource.
```

Author: Updated as suggested.

> * Create PersistentVolume custom resources (CR) for database and file system storage.

Reviewer 2: ```suggestion
* Create persistent volume custom resources (CR) for database and file system storage.
```

Author: Updated as suggested.

> * Install openshift-install CLI for the desired version of {product-title}.

Reviewer 2: What is the `openshift-install` CLI? Do you mean the `oc` CLI? If so:

```suggestion
* You have installed the OpenShift CLI (`oc`).
```

In either case:

> 🚫  desired adjective
Do not use. Use "required" or "the that you want".

Author: Updated as suggested.

> .Next Steps

Reviewer 2: ```suggestion
.Procedure
```

Author: Updated as suggested.

> . Modify `HiveConfig` to enable the feature gate for Assisted Installer:

Reviewer 2: ```suggestion
. Modify the `HiveConfig` resource to enable the feature gate for Assisted Installer:
```

Author: Updated as suggested.

+
[source,terminal]
----

>  oc patch hiveconfig hive --type merge -p '{"spec":{"targetNamespace":"hive","logLevel":"debug","featureGates":{"custom":{"enabled":["AlphaAgentInstallStrategy"]},"featureSet":"Custom"}}}'

Reviewer 2: ```suggestion
$ oc patch hiveconfig hive --type merge -p '{"spec":{"targetNamespace":"hive","logLevel":"debug","featureGates":{"custom":{"enabled":["AlphaAgentInstallStrategy"]},"featureSet":"Custom"}}}'
```

All commands [always](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output) need to start with a $ or #. You need to update them all.

Author: Updated as suggested.

----

. Modify the `provisioning` resource to allow the bare metal operator to watch all namespaces:
+
[source,terminal]
----
 oc patch provisioning provisioning-configuration --type merge -p '{"spec":{"watchAllNamespaces": true }}'
----

> . Create an `AgentServiceConfig` custom resource:

Reviewer 2: You're not actually creating a resource with this command. Does the user need to store the CR in the GitOps repo (that I know exists because I've done a couple of reviews for ZTP content this sprint) or does the user need to do something like create the "custom resource that contains the following YAML:?

Author: Updated as suggested.

+
[source,yaml]
----
 kind: AgentServiceConfig
 metadata:
   name: agent
   namespace: assisted-installer
 spec:
   databaseStorage:
     volumeName: <db_pv_name> <1>
     accessModes:
     - ReadWriteOnce
     resources:
       requests:
         storage: <db_storage_size> <2>
   filesystemStorage:
     volumeName: <fs_pv-name> <3>
     accessModes:
     - ReadWriteOnce
     resources:
       requests:
         storage: <fs_storage_size> <4>
   osImages:
     - openshiftVersion: <ocp_version> <5>
       rootfs: <rootfs_url> <6>
       url: <iso_url> <7>
----

> <1> PersistentVolume name to use for database storage.

Reviewer 2: ```suggestion
<1> The `PersistentVolume` name to use for database storage.
```

This sounds like a reference to a specific PV, so it should be styled. Check out the [object reference](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#general-references) guidelines and update the rest of the annotations.

Author: Updated as suggested.

// Module included in the following assemblies:
//
// scalability_and_performance/ztp-deploying-disconnected.adoc

[id="enabling-assisted-installer-service-on-bare-metal_{context}"]
= Enabling assisted installer service on bare metal

> The Assisted Installer Service (AIS) deploys {product-title} clusters. {rh-rhacm-first} ships with AIS, which is deployed when the MultiClusterHub Operator is enabled on the {rh-rhacm} hub cluster.

Reviewer 2: When this content GAs, I suggest rewriting the introduction so that the user action of the module is in the first paragraph and the context is in a different module.

> For distributed units (DUs), {rh-rhacm} supports {product-title} deployments, which run on a single bare metal host. The single node cluster acts as both a control plane and a worker node. {rh-rhacm} supports single node OpenShift deployments for {product-title} on top of a single bare metal Operator (BMO) DU.

Reviewer: s/single bare metal Operator (BMO) DU./single bare-metal host.

Author: Updated with suggested change.

.Prerequisites

> * Install {product-title} {product-verstion} on a hub cluster.

Reviewer 2: ```suggestion
* Install {product-title} {product-version} on a hub cluster.
```

Reviewer: `{product-version}` is rendering as `Branch Build` in the preview output. @kalexand-rh, do we need to call out product version in the prereqs?

// Module included in the following assemblies:
//
// scalability_and_performance/ztp-deploying-disconnected.adoc

> 

Reviewer 4: Still needs an anchor. [id="enabling-assisted-installer-service-on-bare-metal_{context}"]

Author: Added suggested anchor.

.Enabling Assisted Installer Service on Bare Metal

The Assisted Installer Service (AIS) deploys {product-title} clusters. {rh-rhacm-first} ships with AIS, which is deployed when the MultiClusterHub Operator is enabled on the {rh-rhacm} hub cluster.

> For distributed units DUs, {rh-rhacm} supports single node OpenShift deployments. Single node OpenShift runs on a single bare-metal host. When the host is provisoned, the single node cluster acts as both supervisor and worker node. Shift Container Platform {product-title} on top of a single BMO For distributed units DUs, {rh-rhacm} supports single node OpenShift deployments. Single node OpenShift runs on a single bare-metal host. When the host is provisoned, the single node cluster acts as both supervisor and worker node.

Reviewer 4: First instance of an acronym should be in parens. (DUs).

> For distributed units DUs, {rh-rhacm} supports single node OpenShift deployments. Single node OpenShift runs on a single bare-metal host. When the host is provisoned, the single node cluster acts as both supervisor and worker node. Shift Container Platform {product-title} on top of a single BMO For distributed units DUs, {rh-rhacm} supports single node OpenShift deployments. Single node OpenShift runs on a single bare-metal host. When the host is provisoned, the single node cluster acts as both supervisor and worker node.

Author: Updated as suggested.

> For distributed units DUs, {rh-rhacm} supports single node OpenShift deployments. Single node OpenShift runs on a single bare-metal host. When the host is provisoned, the single node cluster acts as both supervisor and worker node. Shift Container Platform {product-title} on top of a single BMO For distributed units DUs, {rh-rhacm} supports single node OpenShift deployments. Single node OpenShift runs on a single bare-metal host. When the host is provisoned, the single node cluster acts as both supervisor and worker node.

Author: Updated accordingly.

.Enable assisted installer service

.Prerequisites

. Install {product-title} {product-verstion} on a hub cluster.
. Install {rh-acm} and create the MultiClusterHub resource.
. Create PersistentVolume custom resources (CR) for database and file system storage.

> . Install openshift-install CLI for the desired version of {product-title}.

Reviewer 4: This should be an unordered list (bullets), not numbers. * instead of .

Author: Updated accordingly.

.Next Steps

> . Modify `HiveConfig` to enable the feature gate for Assisted Installer

Reviewer 4: Colon at the end.

Author: Updated accordingly.

+
[source,terminal]
----
 oc patch hiveconfig hive --type merge -p '{"spec":{"targetNamespace":"hive","logLevel":"debug","featureGates":{"custom":{"enabled":["AlphaAgentInstallStrategy"]},"featureSet":"Custom"}}}'
----

> . Modify the `provisioning` resource to allow the Bare Metal Operator to watch all namespaces

Reviewer 4: Colon at the end.

Author: Updated accordingly.

+
[source,terminal]
----
 oc patch provisioning provisioning-configuration --type merge -p '{"spec":{"watchAllNamespaces": true }}'
----

. Create an `AgentServiceConfig` custom resource
+
[source,yaml]
----
 kind: AgentServiceConfig
 metadata:
   name: agent
   namespace: assisted-installer
 spec:
   databaseStorage:
     volumeName: <db_pv_name> <1>
     accessModes:
     - ReadWriteOnce
     resources:
       requests:
         storage: <db_storage_size> <2>
   filesystemStorage:
     volumeName: <fs_pv-name> <3>
     accessModes:
     - ReadWriteOnce
     resources:
       requests:
         storage: <fs_storage_size> <4>
   osImages:
     - openshiftVersion: <ocp_version> <5>
       rootfs: <rootfs_url> <6>
       url: <iso_url> <7>
----
<1> PersistentVolume name to use for database storage.
<2> Volume size of the `db_pv_name` PersistentVolume.
<3> PersistentVolume name to use for file system storage.
<4> Volume size of the `fs_pv_name` PersistentVolume.

> <5> {product-title} version to boot the single node OpenShift clusters with.

Reviewer 4: Probably shouldn't end the sentence with a preposition.

Author: Updated accordingly.

<6> The rootfs ISO URL; part of step 4 of the prerequisites.
<7> The image url; part of step 4 of the prerequisites.

> . Install the Assisted Installer Service's controller

Reviewer 4: Colon at the end.

Author: Updated accordingly.

// Module included in the following assemblies:
//
// scalability_and_performance/ztp-deploying-disconnected.adoc

[id="enabling-assisted-installer-service-on-bare-metal_{context}"]

.Enabling Assisted Installer Service on Bare Metal

The Assisted Installer Service (AIS) deploys {product-title} clusters. {rh-rhacm-first} ships with AIS, which is deployed when the MultiClusterHub Operator is enabled on the {rh-rhacm} hub cluster.

> For distributed units DUs, {rh-rhacm} supports single node OpenShift deployments. Single node OpenShift runs on a single bare metal host. When the host is provisoned, the single node cluster acts as both supervisor and worker node. {rh-rhacm} supports single node OpenShift deployments for Shift Container Platform {product-title} on top of a single BMO For distributed units (DUs). Single node OpenShift runs on a single bare-metal host. Once provisoned, the single node cluster acts as both control plane and worker node.

Reviewer 4: First instance of DU is at the beginning of the paragraph. Parens should be there. 

This still sounds odd: {rh-rhacm} supports single node OpenShift deployments for Shift Container Platform {product-title} on top of a single BMO For distributed units (DUs).

What is Shift Container Platform? With {product-title}, that's going to render as "Shift Container Platform OpenShift Container Platform".

Reviewer 4: "Single node OpenShift runs on a single bare-metal host." is a repeat of an earlier sentence. You can delete that.

Author: Updated as suggested.

.Enable assisted installer service

.Prerequisites

* Install {product-title} {product-verstion} on a hub cluster.
* Install {rh-acm} and create the MultiClusterHub resource.
* Create PersistentVolume custom resources (CR) for database and file system storage.
* Install openshift-install CLI for the desired version of {product-title}.

.Next Steps

. Modify `HiveConfig` to enable the feature gate for Assisted Installer:
+
[source,terminal]
----
 oc patch hiveconfig hive --type merge -p '{"spec":{"targetNamespace":"hive","logLevel":"debug","featureGates":{"custom":{"enabled":["AlphaAgentInstallStrategy"]},"featureSet":"Custom"}}}'
----

. Modify the `provisioning` resource to allow the Bare Metal Operator to watch all namespaces:
+
[source,terminal]
----
 oc patch provisioning provisioning-configuration --type merge -p '{"spec":{"watchAllNamespaces": true }}'
----

> . Create an `AgentServiceConfig` custom resource

Reviewer 4: Colon at the end.

Author: Added suggested colon.

+
[source,yaml]
----
 kind: AgentServiceConfig
 metadata:
   name: agent
   namespace: assisted-installer
 spec:
   databaseStorage:
     volumeName: <db_pv_name> <1>
     accessModes:
     - ReadWriteOnce
     resources:
       requests:
         storage: <db_storage_size> <2>
   filesystemStorage:
     volumeName: <fs_pv-name> <3>
     accessModes:
     - ReadWriteOnce
     resources:
       requests:
         storage: <fs_storage_size> <4>
   osImages:
     - openshiftVersion: <ocp_version> <5>
       rootfs: <rootfs_url> <6>
       url: <iso_url> <7>
----
<1> PersistentVolume name to use for database storage.
<2> Volume size of the `db_pv_name` PersistentVolume.
<3> PersistentVolume name to use for file system storage.
<4> Volume size of the `fs_pv_name` PersistentVolume.
<5> {product-title} version to boot the single node OpenShift clusters.

> <6> The rootfs ISO URL; part of step 4 of the prerequisites.

Reviewer 3: Footnote 6 & 7 don't seem accurate to what the prerequisites are above

Author: Removed outdated footnotes as suggested.

<7> The image url; part of step 4 of the prerequisites.

> . Install the Assisted Installer Service's controller:

Reviewer 3: I don't think this step is needed. From my understanding, only the `AgentServiceConfig` needs to be created in order to have the controller be deployed

Author: Updated as suggested.
